### PR TITLE
[FIRRTL] Add type alias parser

### PR DESCRIPTION
**Pull request by uenoku**
Separated from #5167. Add parser changes for type alias. https://github.com/chipsalliance/firrtl-spec/blob/main/spec.md#type-alias

Currently type alias for non-base types are stripped immediately while parsing with warnings.

On top of https://github.com/llvm/circt/pull/5417

### CODE REVIEW

File: test/Dialect/FIRRTL/parse-errors.fir [hunk 1]
```diff
@@ -749,3 +749,8 @@ circuit UnsupportedStringEncodedIntegerLiterals:
     output foo: UInt<8>
     ; expected-error @below {{String-encoded integer literals are unsupported after FIRRTL 3.0.0}}
     connect foo, UInt("h2a")
+
+// -----
+circuit UnknownTypeAlias:
+  module UnknownTypeAlias:
+    output x : Foo ; expected-error {{type identifier `Foo` is not declared}}
```
Comment (prithayan) [inline test/Dialect/FIRRTL/parse-errors.fir:756]: Another test for `if (!getConstants().aliasMap.insert({id, type}).second)
    return emitError(loc) << "type alias " << name << " is already defined";` would be great.

File: lib/Dialect/FIRRTL/Import/FIRParser.cpp [hunk 2]
```diff
@@ -4151,6 +4183,31 @@ ParseResult FIRCircuitParser::parseModule(CircuitOp circuit,
...
+      parseType(type, "expected a type"))
+    return failure();
+  auto name = StringAttr::get(type.getContext(), id);
+  // Create type alias only for base types. Otherwise just pass through the
+  // type.
+  if (auto base = dyn_cast<FIRRTLBaseType>(type))
+    type = BaseTypeAliasType::get(name, base);
+  else
+    emitWarning(loc)
+        << "type alias for non-based type " << type
```
Comment (prithayan) [inline lib/Dialect/FIRRTL/Import/FIRParser.cpp:4203]: Maybe ?
```suggestion
        << "type alias for non-base type " << type
```

**Review by prithayan — APPROVED** [review]

**Comment by dtzSiFive** [comment]
Is there a requirement re:declaring type declarations "before" (in source code position) their use?

Quick testing suggests modules can use type aliases defined after, but type aliases themselves cannot do this.

On one hand, requiring definitions before their use seems reasonable but it does seem a little inconsistent with using other top-level definitions where order doesn't matter.  Is this the desired/expected/preferred behavior?

If nothing else this should be an intentional decision and aligned with spec (didn't check).

**Comment by dtzSiFive** [comment]
Good points re:recursion + needing to figure out how to resolve port types.

Not sure what right answer is, other alternative is require all type aliases def-before-use (and maybe before first module)? Not sure how generator-friendly that would be.

**Comment by uenoku** [comment]
>  other alternative is require all type aliases def-before-use (and maybe before first module)? 

Yep, that sounds reasonable restriction. I agree that it's not bad to limit type declarations to be before first module. Verifying def-before-use is also tricky since we will have to associate type declarations with their parsed positions in fir files. I'm leaning toward just restricting type declarations to be before the first module (as it's simplest).

**Pull request merged**